File: packages/redsys/android/src/main/java/io/capkit/redsys/RedsysImpl.kt
package io.capkit.redsys

import android.app.Activity
import com.redsys.tpvvinapplibrary.IPaymentResult
import com.redsys.tpvvinapplibrary.TPVV
import com.redsys.tpvvinapplibrary.TPVVConfiguration
import com.redsys.tpvvinapplibrary.TPVVConstants
import com.redsys.tpvvinapplibrary.UIDirectPaymentConfig
import io.capkit.redsys.utils.RedsysLogger
import org.json.JSONObject

/**
 * Redsys Native Implementation (Android)
 *
 * This class represents the Implementation Layer of the plugin.
 * It contains pure Android logic and orchestrates the TPVV SDK.
 *
 * Architectural rules:
 * - MUST NOT reference PluginCall
 * - MUST NOT depend on Capacitor bridge APIs
 * - MUST execute UI operations on the main thread
 */
class RedsysImpl(
  private val activity: Activity,
) {
  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Cached immutable plugin configuration.
   * Injected once during plugin load.
   */
  private lateinit var config: RedsysConfig

  // ---------------------------------------------------------------------------
  // Configuration
  // ---------------------------------------------------------------------------

  /**
   * Applies configuration to the native SDK.
   *
   * This method must be called once during plugin initialization.
   */
  fun updateConfig(newConfig: RedsysConfig) {
    this.config = newConfig
    RedsysLogger.verbose = newConfig.verboseLogging
    RedsysLogger.debug(
      "Configuration applied. Verbose logging:",
      newConfig.verboseLogging.toString(),
    )

    // -----------------------------------------------------------------------
    // Global SDK Configuration
    // -----------------------------------------------------------------------

    TPVVConfiguration.setLicense(config.license)
    TPVVConfiguration.setFuc(config.fuc)
    TPVVConfiguration.setTerminal(config.terminal)
    TPVVConfiguration.setCurrency(config.currency)

    // Apply optional merchant configuration
    config.merchantName?.let { TPVVConfiguration.setMerchantName(it) }
    config.merchantUrl?.let { TPVVConfiguration.setMerchantUrl(it) }
    config.titular?.let { TPVVConfiguration.setTitular(it) }

    val env =
      when (config.environment.lowercase()) {
        "real" -> TPVVConstants.ENVIRONMENT_REAL
        "test" -> TPVVConstants.ENVIRONMENT_TEST
        else -> TPVVConstants.ENVIRONMENT_INTEGRATION
      }
    TPVVConfiguration.setEnvironment(env)

    // Apply WebView behavior flags
    TPVVConfiguration.setEnableRedirection(config.enableRedirection)
    TPVVConfiguration.setEnableResultAlert(config.enableResultAlert)

    // Set global language for WebView flows
    config.merchantConsumerLanguage?.let { langCode: String ->
      // Ensure the language code is passed as defined in RedsysLanguage enum
      TPVVConfiguration.setLanguage(langCode)
    }

    // Optional Result Alert Customization (Android SDK 2.4.5)
    if (config.enableResultAlert) {
      // Use configured strings from UI options or fallback to defaults
      TPVVConfiguration.setResultAlertTextOk(
        config.uiResultAlertTextOk ?: "The operation has been completed successfully.",
      )
      TPVVConfiguration.setResultAlertTextKo(
        config.uiResultAlertTextKo ?: "An error occurred while processing the operation.",
      )
      TPVVConfiguration.setResultAlertTextButtonOk(config.uiResultAlertButtonTextOk ?: "Continue")
      TPVVConfiguration.setResultAlertTextButtonKo(config.uiResultAlertButtonTextKo ?: "Continue")
    }

    // Reset Direct Payment UI static state before applying new values
    UIDirectPaymentConfig.setTopBarColor(null)
    UIDirectPaymentConfig.setBackgroundColor(null)
    UIDirectPaymentConfig.setProgressBarColor(null)
    UIDirectPaymentConfig.setBtnText(null)

    // Apply Direct Payment UI customization using static SDK configuration
    val uiConfig = UIDirectPaymentConfig()

    // Apply granular UI settings found in UIDirectPaymentConfig.class
    config.uiCardHeaderBgColor?.let { UIDirectPaymentConfig.setCardHeadTextBackgroundColor(it) }
    config.uiCardHeaderText?.let { UIDirectPaymentConfig.setCardHeadText(it) }
    config.uiLabelTextColor?.let {
      UIDirectPaymentConfig.setCardHeadTextColor(it)
      UIDirectPaymentConfig.setInfoTextColor(it)
    }

    // Global progress bar color
    config.uiProgressBarColor?.let { TPVVConfiguration.setProgressBarColor(it) }

    TPVVConfiguration.setUiDirectPaymentConfig(uiConfig)
  }

  // ---------------------------------------------------------------------------
  // Direct Payment
  // ---------------------------------------------------------------------------

  /**
   * Executes the direct payment flow.
   * Resets and applies UI static configuration for each transaction to emulate instance behavior.
   */
  fun doDirectPayment(
    order: String,
    amount: Double,
    type: String,
    id: String?,
    desc: String?,
    extra: HashMap<String, String>?,
    uiOptions: JSONObject?,
    callback: IPaymentResult,
  ) {
    activity.runOnUiThread {
      // 1. Reset static SDK UI state to defaults
      UIDirectPaymentConfig.setTopBarColor(null)
      UIDirectPaymentConfig.setBackgroundColor(null)
      UIDirectPaymentConfig.setProgressBarColor(null)
      UIDirectPaymentConfig.setBtnText(null)
      UIDirectPaymentConfig.setCardHeadTextBackgroundColor(null)

      // 2. Apply runtime UI overrides from JavaScript if present
      uiOptions?.let { ui ->
        ui.getString("backgroundColor")?.let { UIDirectPaymentConfig.setBackgroundColor(it) }
        ui.getString("confirmButtonText")?.let { UIDirectPaymentConfig.setBtnText(it) }
        ui.getString("androidTopBarColor")?.let { UIDirectPaymentConfig.setTopBarColor(it) }
        ui.getString("cardHeaderBackgroundColor")?.let { UIDirectPaymentConfig.setCardHeadTextBackgroundColor(it) }
      }

      TPVV.doDirectPayment(activity, order, amount, type, id, desc, extra, callback)
    }
  }

  // ---------------------------------------------------------------------------
  // Web Payment Initialization (Phase 1)
  // ---------------------------------------------------------------------------

  /**
   * Initializes the WebView payment flow.
   * Note: The SDK method name contains a typo 'inicializeWebViewPayment'.
   */
  fun initializeWebPayment(
    order: String,
    amount: Double,
    type: String,
    id: String?,
    desc: String?,
    extra: HashMap<String, String>?,
    onResult: (String) -> Unit,
    onError: (RedsysError) -> Unit,
  ) {
    activity.runOnUiThread {
      try {
        // We use the mandatory 'inicializeWebViewPayment' name from the SDK binary.
        TPVV.inicializeWebViewPayment(
          activity,
          order,
          amount,
          type,
          id ?: "", // Ensure non-null string for identifier
          desc ?: "",
          extra,
          object : com.redsys.tpvvinapplibrary.webviewPayment.IPaymentWVResult {
            override fun paymentResultOK(base64Data: String?) {
              if (base64Data != null) {
                onResult(base64Data)
              } else {
                onError(RedsysError.InitFailed("Base64 data is null"))
              }
            }

            override fun paymentResultKO(error: com.redsys.tpvvinapplibrary.ErrorResponse?) {
              onError(RedsysError.InitFailed(error?.desc ?: "Web payment initialization failed"))
            }
          },
        )
      } catch (e: Exception) {
        onError(RedsysError.InitFailed("Exception during WebView initialization: ${e.message}"))
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Web Payment Execution (Phase 2)
  // ---------------------------------------------------------------------------

  /**
   * Executes the WebView payment flow.
   *
   * The SDK handles Activity presentation internally.
   */
  fun processWebPayment(
    signature: String,
    callback: IPaymentResult,
  ) {
    activity.runOnUiThread {
      TPVV.doWebPayment(signature, activity, callback)
    }
  }
}

File: packages/redsys/ios/Sources/RedsysPlugin/RedsysImpl.swift
import Foundation
import TPVVInLibrary

/**
 * Redsys Native Implementation (iOS)
 *
 * This class represents the Implementation Layer of the plugin.
 * It contains platform-specific logic and orchestrates the TPVVInLibrary SDK.
 *
 * Architectural rules:
 * - MUST NOT access CAPPluginCall
 * - MUST NOT depend on Capacitor APIs
 * - MUST return platform-agnostic results to the Bridge layer
 */
@objc public final class RedsysImpl: NSObject {

    // MARK: - Properties

    /// Cached immutable configuration injected from the Bridge layer.
    private var config: RedsysConfig?

    // MARK: - Initialization

    /**
     * Initializes the implementation instance.
     */
    override init() {
        super.init()
    }

    // MARK: - Configuration

    /**
     Applies plugin configuration and initializes the TPVV SDK.

     This method must be called exactly once during plugin load.
     */
    func applyConfig(_ config: RedsysConfig) {

        // Ensure configuration is applied only once.
        precondition(
            self.config == nil,
            "RedsysImpl.applyConfig(_:) must be called exactly once"
        )
        self.config = config

        // Synchronize logger state
        RedsysLogger.verbose = config.verboseLogging
        RedsysLogger.debug("Configuration applied. Verbose logging:", config.verboseLogging)

        // MARK: - SDK Global Configuration

        TPVVConfiguration.shared.appLicense = config.license
        TPVVConfiguration.shared.appFuc = config.fuc
        TPVVConfiguration.shared.appTerminal = config.terminal
        TPVVConfiguration.shared.appCurrency = config.currency

        // Apply optional merchant configuration
        if let titular = config.titular {
            TPVVConfiguration.shared.appMerchantTitular = titular
        }

        if let merchantName = config.merchantName {
            TPVVConfiguration.shared.appMerchantName = merchantName
        }

        if let merchantUrl = config.merchantUrl {
            TPVVConfiguration.shared.appMerchantURL = merchantUrl
        }

        // Configure redirection URLs for 3DS flow
        // Note: If set, the user must manually tap 'Back' to close the WebView
        if let urlOK = config.merchantUrl { // Usually merchantUrl acts as base
            TPVVConfiguration.shared.appURLOK = urlOK
        }
        if let urlKO = config.merchantUrl {
            TPVVConfiguration.shared.appURLKO = urlKO
        }

        // These fields may not always be present but are supported by SDK
        if let merchantDescription = config.merchantDescription {
            TPVVConfiguration.shared.appMerchantDescription = merchantDescription
        }

        if let merchantData = config.merchantData {
            TPVVConfiguration.shared.appMerchantData = merchantData
        }

        if let merchantGroup = config.merchantGroup {
            TPVVConfiguration.shared.appMerchantGroup = merchantGroup
        }

        // MARK: - Environment Mapping

        switch config.environment.lowercased() {
        case "real": TPVVConfiguration.shared.appEnviroment = .Real
        case "test": TPVVConfiguration.shared.appEnviroment = .Test
        default: TPVVConfiguration.shared.appEnviroment = .Integration
        }

        RedsysLogger.debug("RedsysImpl: SDK initialized in \(TPVVConfiguration.shared.appEnviroment) mode.")

        // Apply WebView payment methods
        if let methods = config.paymentMethods {
            TPVVConfiguration.shared.appMerchantPayMethods =
                RedsysUtils.mapPaymentMethods(methods).first ?? PaymentMethod.card
        }

        // Apply WebView language
        // Note: This setting affects only the WebView/3DS flow.
        if let language = config.merchantConsumerLanguage {
            TPVVConfiguration.shared.appMerchantConsumerLanguage = language
        }
    }

    // MARK: - Web Payment (Phase 1: Initialization)

    /**
     * Initializes the WebView payment flow on iOS.
     */
    @objc func initializeWebPayment(
        order: String,
        amount: Double, // Using Double as per swiftinterface inspection
        type: TransactionType,
        description: String?,
        identifier: String?,
        extraParams: [String: String]?,
        completion: @escaping (String?, Error?) -> Void
    ) {
        // SDK_INAPP.inicializeWebPayment is the official method in the framework.
        SDK_INAPP.inicializeWebPayment(
            orderNumber: order,
            amount: amount,
            productDescription: description ?? "",
            transactionType: type,
            identifier: identifier ?? "",
            extraParams: extraParams ?? [:],
            success: { base64String in
                completion(base64String, nil)
            },
            failure: { error in
                completion(nil, RedsysError.initFailed(error.desc))
            }
        )
    }

    // MARK: - Web Payment (Phase 2: Execution)

    /**
     Executes the WebView payment flow (Step 2).

     Presents the WebView controller and assigns the delegate
     for asynchronous result handling.
     */
    @objc func processWebPayment(
        signature: String,
        signatureVersion: String?,
        delegate: WebViewPaymentResponseDelegate
    ) {
        DispatchQueue.main.async {
            let webVC = SDK_INAPP.doWebViewPayment(
                signature: signature,
                signatureVersion: signatureVersion ?? "HMAC_SHA256_V1"
            )
            webVC.delegate = delegate

            // Retrieve active scene rootViewController in a scene-safe way
            if let windowScene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first,
               let rootVC = windowScene.windows
                .first(where: { $0.isKeyWindow })?
                .rootViewController {

                rootVC.present(webVC, animated: true)
            }
        }
    }

    // MARK: - Direct Payment

    /**
     Executes the direct payment flow.

     Presents the native DirectPaymentViewController and
     assigns the delegate for asynchronous result handling.
     */
    @objc func executeDirectPayment(
        order: String,
        amount: Double,
        type: TransactionType,
        description: String?,
        identifier: String?,
        extraParams: [String: String]?,
        uiOptions: [String: Any]?,
        delegate: DirectPaymentResponseDelegate
    ) {
        DispatchQueue.main.async {
            let uiConfig = TPVInAppUIConfig()

            // Retrieve label color from config or default to black
            let labelColor = RedsysUtils.colorFromHex(self.config?.labelTextColor ?? "#000000") //
            let labelFont = UIFont.systemFont(ofSize: 16)

            // Background & Buttons using _ to ignore the returned value from fluent API
            if let bgColorHex = self.config?.uiBackgroundColor {
                _ = uiConfig.setBackgorundViewColor(color: RedsysUtils.colorFromHex(bgColorHex))
            }

            // Apply Runtime Overrides from uiOptions (JS Call)
            if let ui = uiOptions {
                if let bgColor = ui["backgroundColor"] as? String {
                    _ = uiConfig.setBackgorundViewColor(color: RedsysUtils.colorFromHex(bgColor))
                }
                if let btnText = ui["confirmButtonText"] as? String {
                    _ = uiConfig.setContinueButtonText(btnText)
                }
            }

            if let btnText = self.config?.uiConfirmButtonText { _ = uiConfig.setContinueButtonText(btnText) }
            if let cancelText = self.config?.uiCancelButtonText { _ = uiConfig.setCancelButtonText(cancelText) }

            // Labels customization
            if let cardLabelText = self.config?.cardNumberLabel {
                let labelCfg = PaymentUILabel.create(text: cardLabelText, textColor: labelColor, textFont: labelFont)
                _ = uiConfig.setNumberCardLabel(labelCfg)
            }
            if let expiryLabelText = self.config?.expirationLabel {
                let labelCfg = PaymentUILabel.create(text: expiryLabelText, textColor: labelColor, textFont: labelFont)
                _ = uiConfig.setExpireDateLabel(labelCfg)
            }
            if let cvvLabelText = self.config?.cvvLabel {
                let labelCfg = PaymentUILabel.create(text: cvvLabelText, textColor: labelColor, textFont: labelFont)
                _ = uiConfig.setCVVLabel(labelCfg)
            }

            // Logo from assets
            if let logoName = self.config?.uiLogo, let logoImage = UIImage(named: logoName) {
                _ = uiConfig.setLogo(logo: logoImage)
            }

            let paymentVC = DirectPaymentViewController(
                orderNumber: order,
                amount: amount,
                productDescription: description ?? "",
                transactionType: type,
                identifier: identifier ?? "",
                extraParams: extraParams ?? [:],
                uiViewConfig: uiConfig
            )

            paymentVC.delegate = delegate

            // Retrieve active scene rootViewController in a scene-safe way
            if let windowScene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first,
               let rootVC = windowScene.windows
                .first(where: { $0.isKeyWindow })?
                .rootViewController {

                rootVC.present(paymentVC, animated: true)
            }
        }
    }
}
